AmplifyPlugins/Geo/Sources/AWSLocationGeoPlugin/Dependency/AWSDeviceTracker.swift

```diff
@@ -35,31 +39,222 @@ class AWSDeviceTracker: NSObject, CLLocationManagerDelegate, DeviceTrackingBehav
+                    } )
+                    locationsToSend.append(contentsOf: locationsReceived)
+
+                    didUpdateLocations(locationsToSend)
+                } catch {
+                    // TODO: send error on Hub
+                }
+            } else {
+                // send to AWS Location
+                Task {
+                    do {
+                        try await sendStoredLocationsToService()
+                        try await sendReceivedLocationsToService(locations: locations,
+                                                                 options: Geo.UpdateLocationOptions(tracker: options.tracker!))
+                    } catch {
+                        // TODO: send error on Hub
+                    }
+                }
+            }
+        }  else {
+            // batch save the locations to local store
+            guard let deviceId = UserDefaults.standard.object(forKey: AWSDeviceTracker.deviceIDKey) as? String else {
+                Amplify.log.error("Not able to fetch deviceId from UserDefaults")
+                return
+            }
+
+            let positions = locations.map({ PositionInternal(timeStamp: currentTime,
+                                                             latitude: $0.coordinate.latitude,
+                                                             longitude: $0.coordinate.longitude,
+                                                             tracker: options.tracker!,
+                                                             deviceID: deviceId) })
+            do {
+                try self.locationStore.insert(positions: positions)
+            } catch {
+                // TODO: send error on Hub
+            }
+        }
+        
+        // save lastLocation and last time a location update is received
+        UserDefaults.standard.set(currentTime, forKey: AWSDeviceTracker.lastLocationUpdateTimeKey)
+        if let lastReceivedLocation = locations.last,
+            let encodedLocation = try? NSKeyedArchiver.archivedData(withRootObject: lastReceivedLocation, requiringSecureCoding: false) {
+            UserDefaults.standard.set(encodedLocation, forKey: AWSDeviceTracker.lastUpdatedLocationKey)
+        }
+    }
+    
+    func getLocationsFromLocalStore() throws -> [PositionInternal] {
+        let storedLocations = try locationStore.getAll()
+        try locationStore.removeAll()
+        return storedLocations
+    }
+    
+    func sendReceivedLocationsToService(locations: [CLLocation], options: Geo.UpdateLocationOptions) async throws {
+        guard let deviceId = UserDefaults.standard.object(forKey: AWSDeviceTracker.deviceIDKey) as? String else {
+            Amplify.log.error("Not able to fetch deviceId from UserDefaults")
+            return
+        }
+        
+        for location in locations {
+            try await Amplify.Geo.updateLocation(Geo.Location(clLocation: location.coordinate),
```

Comment:
can this be optimized to go directly to the service api and batch each request with 10 locations?